File: org.nickvision.application.gnome/src/views/preferencesdialog.cpp
#include "views/preferencesdialog.h"
#include "helpers/gtkhelpers.h"

using namespace Nickvision::Application::Shared::Controllers;
using namespace Nickvision::Application::Shared::Models;
using namespace Nickvision::Application::GNOME::Helpers;
using namespace Nickvision::Events;

namespace Nickvision::Application::GNOME::Views
{
    PreferencesDialog::PreferencesDialog(const std::shared_ptr<PreferencesViewController>& controller, GtkWindow* parent)
        : DialogBase{ parent, "preferences_dialog" },
        m_controller{ controller }
    {
        //Load
        adw_combo_row_set_selected(m_builder.get<AdwComboRow>("themeRow"), static_cast<unsigned int>(m_controller->getTheme()));
        GtkHelpers::setComboRowModel(m_builder.get<AdwComboRow>("languageRow"), m_controller->getAvailableTranslationLanguages(), m_controller->getTranslationLanguage());
        //Signals
        m_closed += [&](const EventArgs&){ onClosed(); };
        g_signal_connect(m_builder.get<GObject>("themeRow"), "notify::selected-item", G_CALLBACK(+[](GObject*, GParamSpec* pspec, gpointer data){ reinterpret_cast<PreferencesDialog*>(data)->onThemeChanged(); }), this);
    }

    void PreferencesDialog::onClosed()
    {
        m_controller->setTranslationLanguage(static_cast<size_t>(adw_combo_row_get_selected(m_builder.get<AdwComboRow>("languageRow"))));
        m_controller->saveConfiguration();
    }

    void PreferencesDialog::onThemeChanged()
    {
        m_controller->setTheme(static_cast<Theme>(adw_combo_row_get_selected(m_builder.get<AdwComboRow>("themeRow"))));
        switch (m_controller->getTheme())
        {
        case Theme::Light:
            adw_style_manager_set_color_scheme(adw_style_manager_get_default(), ADW_COLOR_SCHEME_FORCE_LIGHT);
            break;
        case Theme::Dark:
            adw_style_manager_set_color_scheme(adw_style_manager_get_default(), ADW_COLOR_SCHEME_FORCE_DARK);
            break;
        default:
            adw_style_manager_set_color_scheme(adw_style_manager_get_default(), ADW_COLOR_SCHEME_DEFAULT);
            break;
        }
    }
}

File: Nickvision.Application.GNOME/Views/PreferencesDialog.cs
using Nickvision.Application.Shared.Controllers;
using Nickvision.Application.Shared.Models;
using Nickvision.Desktop.GNOME.Helpers;
using System;

namespace Nickvision.Application.GNOME.Views;

public class PreferencesDialog : Adw.PreferencesDialog
{
    private readonly PreferencesViewController _controller;
    private readonly Gtk.Builder _builder;

    [Gtk.Connect("themeRow")]
    private Adw.ComboRow? _themeRow;
    [Gtk.Connect("languageRow")]
    private Adw.ComboRow? _languageRow;

    public PreferencesDialog(PreferencesViewController controller) : this(controller, Gtk.Builder.NewFromBlueprint("PreferencesDialog", controller.Translator))
    {

    }

    private PreferencesDialog(PreferencesViewController controller, Gtk.Builder builder) : base(new Adw.Internal.PreferencesDialogHandle(builder.GetPointer("root"), false))
    {
        _controller = controller;
        _builder = builder;
        _builder.Connect(this);
        // Load
        _themeRow!.SetModel(_controller.Themes);
        _languageRow!.SetModel(_controller.AvailableTranslationLanguages);
        // Events
        OnClosed += Dialog_OnClosed;
        _themeRow!.OnNotify += ThemeRow_OnNotify;
    }

    private async void Dialog_OnClosed(Adw.Dialog sender, EventArgs args)
    {
        _controller.TranslationLanguage = _controller.AvailableTranslationLanguages[(int)_languageRow!.Selected];
        await _controller.SaveConfigurationAsync();
    }

    private void ThemeRow_OnNotify(GObject.Object sender, NotifySignalArgs args)
    {
        if (args.Pspec.GetName() == "selected-item")
        {
            _controller.Theme = _controller.Themes[(int)_themeRow!.Selected];
            Adw.StyleManager.GetDefault().ColorScheme = _controller.Themes[(int)_themeRow!.Selected].Value switch
            {
                Theme.Light => Adw.ColorScheme.ForceLight,
                Theme.Dark => Adw.ColorScheme.ForceDark,
                _ => Adw.ColorScheme.Default
            };
        }
    }
}
